### DISCUSSION TITLE
Compiling a Next.JS library

### DISCUSSION THREAD

**Discussion by josias-r** [vercel/next.js#13399] [Help]
I'm working on an NPM library based on NextJS. When I will publish it, it's supposed to be a library where a developer can import Components and utils from.

Now I'm wondering what the most convenient set-up is for compilation.
I'm asking because I always find my self having to tune the compilation steps. I ended up w/ simply using the NextJS babel preset for a long time, but once again I have troubles because dynamic imports don't work anymore after they have been compiled.

So I tried other things like using the preset-env together w/ the react-preset w/ using the NextJS babel preset. That introduced way more errors.

Bottom line: What is the cleanest way where I won't lose functionality through compilation like in my example above?

Maybe it would be easier to simply whitelist it the webpack config? But I want to avoid that because that would be an extra step for any new user trying to install the package.

I would love to see something like a babel preset that would be exactly for creating external libraries, that would relieve from all the compilation hustles.

**Comment by alexisvisco** [comment]
Same issue here with the library of a friend https://github.com/antoineaudrain/react-editorjs-renderer.

How to do a compatible nextjs library ?

**Reply by josias-r** [reply]
@alexisvisco I currently went with the solution to use `const withTM = require("next-transpile-modules")["my-package"]` from the next config. This way your package will get built once the nextjs builts and you don't need to build your package yourself first.
It works really well.

**Reply by alexisvisco** [reply]
Hello, your solution seems not resolving our problem. Are you transpiling with babel or rollup for your library?
Can you maybe share your library with me so I can give it a look?

**Reply by josias-r** [reply]
Hello, sorry the library is currently private, but here's my config:

Inside the consumer nextjs project:
```
const withPlugins = require("next-compose-plugins");
const withTM = require("next-transpile-modules")(["@your/external-package"]);

module.exports = withPlugins(
  [
    [withTM],
  ],
  {
    target: "serverless",
    pageExtensions: ["ts", "tsx"],
    webpack(config, { dev }) {
        ....
    },
  }
);
```

Inside the package itself
I do not use babel, but I do use TypeScript, here's my full tsconfig from the package (build script: `cross-env NODE_ENV=production npm run tsc`):

```
{
  "compilerOptions": {
    "target": "esnext",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "declaration": true,
    "declarationMap": true,
    "jsx": "preserve",
    "outDir": "dist",

    "baseUrl": "src",
    "plugins": [
      {
        "transform": "@zerollup/ts-transform-paths",
        "exclude": ["*"]
      }
    ]
  },
  "ts-node": {
    "compilerOptions": {
      "target": "es2018",
      "module": "commonjs"
    }
  },
  "include": ["next-env.d.ts", "src"],
  "exclude": ["node_modules"]
}

```
